test/integration/variable_test.rb

```diff
@@ -80,10 +80,15 @@ def test_hash_with_default_proc
     assigns['test'] = 'Tobi'
     assert_equal 'Hello Tobi', template.render!(assigns)
     assigns.delete('test')
-    e = assert_raises(RuntimeError) do
-      template.render!(assigns)
-    end
-    assert_equal "Unknown variable 'test'", e.message
+    assert_equal "Hello ", template.render!(assigns)
+  end
+
+  def test_environment_falsy
+    template = Template.parse(%({{ test }}{% assign test = 'bar' %}{{ test }}))
+    template.assigns['test'] = 'foo'
+    assert_equal 'foobar', template.render!
+    assert_equal 'bar', template.render!('test' => nil)
```

Comment:
Before these changes this would have previously failed and been `foobar`

Reply:
So, is this PR a bug-fix or a breaking-chamge?

Reply:
I believe this is unexpected behaviour that developed from very small changes over a long period. It is very rare that anyone would by relying on this as it requires using `template.assign` and then passing a similar hash to the render method with a nil value. 

It has created a fork in how scopes and environments work and I'm trying to stop that from continuing. They should always be treated the same.

Reply:
If this was picked up sooner I would have called it a bug-fix, the fact this has been in the code so long I call it a breaking change.

Reply:
Maybe I would still call it a bug fix, it just hasn't caused anyone any headaches that they know about

Reply:
This is kind of like https://github.com/Shopify/liquid/issues/345 and is part of where this original code came from but instead of fixing false values it also fixes nil